OpenRA.Mods.Common/Traits/Upgrades/UpgradeManager.cs

```diff
@@ -66,17 +85,36 @@ public UpgradeManager(ActorInitializer init)
 			});
 		}
 
-		public void GrantTimedUpgrade(Actor self, string upgrade, int duration)
+		/// <summary>Upgrade level increments are limited to one per source, i.e., if a single source
+		/// attempts granting multiple upgrades, they will not accumulate. They will replace each other
+		/// instead, leaving only the last granted upgrade active. An upgrade from each new distinct
+		/// source will increment the upgrade's level until AcceptsUpgrade starts returns false. Then,
+		/// when no new levels are accepted, the upgrade source with the shortest remaining upgrade
+		/// duration will be replaced by the new source.</summary>
+		public void GrantTimedUpgrade(Actor self, string upgrade, int duration, object source = null)
 		{
 			var timed = timedUpgrades.FirstOrDefault(u => u.Upgrade == upgrade);
 			if (timed == null)
 			{
-				timed = new TimedUpgrade(upgrade, duration);
+				timed = new TimedUpgrade(upgrade, duration, source);
 				timedUpgrades.Add(timed);
 				GrantUpgrade(self, upgrade, timed);
+				return;
+			}
+
+			var src = timed.Sources.FirstOrDefault(s => s.Source == source);
+			if (source == null || src == null)
+			{
+				timed.Sources.Add(new TimedUpgrade.UpgradeSource(duration, source));
+				if (AcceptsUpgrade(self, upgrade))
+					GrantUpgrade(self, upgrade, timed);
+				else
+					timed.Sources.Remove(timed.Sources.OrderByDescending(s => s.Remaining).Last());
```

Comment:
@GraionDilach, in that case this code doesn't touch the upgrades (neither `GrantUpgrade` nor `RevokeUpgrade` get called). Notice that `GrantUpgrade` and `RevokeUpgrade` use `TimedUpgrade timed` as the upgrade source which grants a certain number of upgrades. And `timed` handles its own housekeeping (sources and remaining time for individual upgrades) separately.
This means that `GrantTimedUpgrade` just updates the information on to whom the upgrades are related. Instead of the two granted upgrades being related to A and B, they will be updated to relate to B and C.